app/models/upload.rb

```diff
@@ -316,6 +317,71 @@ def thumbnail_height
     get_dimension(:thumbnail_height)
   end
 
+  def dominant_color(calculate_if_missing: false)
+    val = read_attribute(:dominant_color)
+    if val.nil? && calculate_if_missing
+      calculate_dominant_color!
+      read_attribute(:dominant_color)
+    else
+      val
+    end
+  end
+
+  def calculate_dominant_color!(local_path = nil)
+    color = nil
+
+    if !FileHelper.is_supported_image?("image.#{extension}") || extension == "svg"
+      color = ""
+    end
+
+    if color.nil?
+      local_path ||=
+        if local?
+          Discourse.store.path_for(self)
+        else
+          Discourse.store.download(self).path
+        end
+
+      color = begin
+        data = Discourse::Utils.execute_command(
+          "nice",
+          "-n",
+          "10",
+          "convert",
+          local_path,
+          "-resize",
+          "1x1",
+          "-define",
+          "histogram:unique-colors=true",
+          "-format",
+          "%c",
+          "histogram:info:",
+          timeout: DOMINANT_COLOR_COMMAND_TIMEOUT_SECONDS
+        )
+
+        # Output format:
+        # 1: (110.873,116.226,93.8821) #6F745E srgb(43.4798%,45.5789%,36.8165%)
+
+        color = data[/#([0-9A-F]{6})/, 1]
+
+        raise "Calculated dominant color but unable to parse output:\n#{data}" if color.nil?
```

Comment:
Raising here feels dangerous though since we might end up blowing up the periodic job. Previously, we would rescue and log this error and I think we should continue to do that.

Reply:
This raise should never happen - it indicates a problem with imagemagick and/or the parsing code. So I think I'd prefer to have it raise all the way up and blow up the job to make sure we notice and fix the problem.

The silently-rescued errors can be caused by bad user input, so that's why I've removed the logging there. If there's no action to take based on an error log, I figure it shouldn't have been an error log in the first place.